packages/telemetry/src/context-aware-slog.js

```diff
@@ -0,0 +1,348 @@
+        break;
+      }
+      case SLOG_TYPES.REPLAY.START: {
+        replayContext = { replay: true };
+        break;
+      }
+      case SLOG_TYPES.DELIVER: {
+        if (replayContext) {
+          assert(finalBody.replay);
+          replayContext = {
+            ...replayContext,
+            'crank.vatID': finalBody.vatID,
+            'crank.deliveryNum': finalBody.deliveryNum,
+          };
+        } else {
+          assert(!!crankContext && !finalBody.replay);
+          crankContext = {
+            ...crankContext,
+            'crank.vatID': finalBody.vatID,
+            'crank.deliveryNum': finalBody.deliveryNum,
+          };
+        }
+
+        delete finalBody.deliveryNum;
+        delete finalBody.replay;
+
+        break;
+      }
+      case SLOG_TYPES.DELIVER_RESULT: {
+        delete finalBody.deliveryNum;
+        delete finalBody.replay;
+
+        break;
+      }
+      case SLOG_TYPES.SYSCALL:
+      case SLOG_TYPES.SYSCALL_RESULT: {
+        eventLogAttributes['crank.syscallNum'] = finalBody.syscallNum;
+
+        delete finalBody.deliveryNum;
+        delete finalBody.replay;
+        delete finalBody.syscallNum;
+
+        break;
+      }
+      case SLOG_TYPES.CONSOLE: {
+        delete finalBody.crankNum;
+        delete finalBody.deliveryNum;
+
+        break;
+      }
+      default:
+        // All other log types are logged as is (using existing contexts) without
+        // any change to the slogs or any contributions to the contexts. This also
+        // means that any unexpected slog type will pass through. To fix that, add
+        // all remaining cases of expected slog types above with a simple break
+        // statement and log a warning here
+        break;
+    }
+
+    const logAttributes = {
+      'process.uptime': monotime,
+      ...initContext, // Optional prelude
+      ...blockContext, // Block is the first level of execution nesting
+      ...triggerContext, // run and trigger info is nested next
+      ...crankContext, // Finally cranks are the last level of nesting
+      ...replayContext, // Replay is a substitute for crank context during vat page in
+      ...eventLogAttributes,
+      ...staticContext,
+    };
+
+    /**
+     * Add any after report operations here
+     * like resetting context data
+     */
+    switch (body.type) {
+      case SLOG_TYPES.KERNEL.INIT.FINISH: {
+        initContext = null;
+        break;
+      }
+      case SLOG_TYPES.COSMIC_SWINGSET.END_BLOCK.START: {
+        triggerContext = restoreContext();
+        break;
+      }
+      case SLOG_TYPES.COSMIC_SWINGSET.AFTER_COMMIT_STATS: {
+        blockContext = null;
+        break;
+      }
+      case SLOG_TYPES.COSMIC_SWINGSET.BOOTSTRAP_BLOCK.FINISH: {
+        blockContext = null;
+        break;
+      }
+      // eslint-disable-next-line no-restricted-syntax
+      case SLOG_TYPES.COSMIC_SWINGSET.RUN.FINISH: {
+        triggerContext = null;
```

Comment:
I made a mistake on the logic of the trigger context persistence. We need to conditionally persist the context here:

```suggestion
        persistContext(finalBody.remainingBeans > 0 ? null : triggerContext);
        triggerContext = null;
```

Then we can simply restore the context on run start for `runNum === 0` I think.

Reply:
Actually I think we can persist an empty context instead (so you can keep `null` as missing).

```suggestion
        persistContext(finalBody.remainingBeans > 0 ? {} : triggerContext);
        triggerContext = null;
```

Reply:
I think we should also add an assertion here for `triggerContext` to exist here right?

Reply:
Since we have no assertion on `triggerContext` here, it could be `null` at this point. So I think we should either do
```
persistContext(finalBody.remainingBeans || !triggerContext ? {} : triggerContext);
```
Or have an assertion:
```
assert(!!triggerContext);
persistContext(finalBody.remainingBeans ? {} : triggerContext)
```

Reply:
There should be a trigger context by this point. Let's add an assertion

Reply:
Done

Reply:
I just realized that in the final version the `> 0` part of the check was dropped, which causes the persisted context to always be empty.

Reply:
So we are missing the context in case of negative remaining beans right?

Reply:
We are missing the persisted context roughly always because the remaining beans is almost never 0.